packages/contracts-bedrock/test/L1/ProxyAdminOwnedBase.t.sol

```diff
@@ -2,67 +2,257 @@
+        // Create a new harness
+        harness = new ProxyAdminOwnedBase_Harness();
 
+        // Set the owner of the harness to the ProxyAdmin contract.
         vm.store(
-            address(proxyAdminOwned),
-            bytes32(Constants.PROXY_OWNER_ADDRESS),
-            bytes32(uint256(uint160(address(proxyAdmin))))
+            address(harness), bytes32(Constants.PROXY_OWNER_ADDRESS), bytes32(uint256(uint160(address(proxyAdmin))))
         );
     }
+}
 
-    function _mockAndExpect(address _target, bytes memory _call, bytes memory _return) internal {
-        vm.mockCall(_target, _call, _return);
-        vm.expectCall(_target, _call);
+contract ProxyAdminOwnedBase_proxyAdminOwner_Test is ProxyAdminOwnedBase_TestInit {
+    /// @notice Tests that the proxyAdminOwner function returns the correct owner.
+    function test_proxyAdminOwner_succeeds() public view {
+        assertEq(harness.proxyAdminOwner(), proxyAdminOwner);
     }
+}
 
-    // Test that the `proxyAdminOwner` function returns the correct owner
-    function test_proxyAdminOwner_succeeds() public {
-        vm.expectCall(address(proxyAdmin), abi.encodeCall(IOwnable.owner, ()));
-        assertEq(proxyAdminOwned.proxyAdminOwner(), owner);
+contract ProxyAdminOwnedBase_proxyAdmin_Test is ProxyAdminOwnedBase_TestInit {
+    /// @notice Tests that the proxyAdmin function returns the correct proxy.
+    function test_proxyAdmin_succeeds() public view {
+        assertEq(address(harness.proxyAdmin()), address(proxyAdmin));
     }
 
-    // Test that the `_sameProxyAdminOwner` function returns true if the proxy admin owner owner
-    // of both proxies is the same
-    function test_sameProxyAdminOwner_sameOwner_succeeds(address _proxy) public {
+    /// @notice Tests that the proxyAdmin function returns the correct proxy when the current
+    ///         contract is a full ResolvedDelegateProxy.
+    function test_proxyAdmin_fullResolvedDelegateProxy_succeeds() public {
+        // Unset the standard proxy owner slot.
+        vm.store(address(harness), bytes32(Constants.PROXY_OWNER_ADDRESS), bytes32(0));
+
+        // Store the string "OVM_L1CrossDomainMessenger" in slot 0.
+        harness.setSlot0(address(harness), "OVM_L1CrossDomainMessenger");
+
+        // Store the address of the proxyAdmin in slot 1.
+        harness.setSlot1(address(harness), address(addressManager));
+
+        // Expect no revert.
+        assertEq(address(harness.proxyAdmin()), address(proxyAdmin));
+    }
+
+    /// @notice Tests that the proxyAdmin function reverts if the current contract is not a
+    ///         ResolvedDelegateProxy.
+    /// @param _slot0Value The value to store in slot 0.
+    function test_proxyAdmin_notResolvedDelegateProxy_reverts(string memory _slot0Value) public {
+        // Assume the slot 0 value is not "OVM_L1CrossDomainMessenger".
+        vm.assume(keccak256(abi.encode(_slot0Value)) != keccak256(abi.encode("OVM_L1CrossDomainMessenger")));
```

Comment:
Is this right? 
In the ResolvedDelegateProxy the mapping is: 

https://github.com/ethereum-optimism/optimism/blob/b00761795084c248100b441fa5e32ca746785d1f/packages/contracts-bedrock/src/legacy/ResolvedDelegateProxy.sol#L18

So I don't think the `_slot0Value` would just be `keccak256(abi.encode("OVM_L1CrossDomainMessenger"))`?

Reply:
I responded to myself at the same time. Check [this response](https://github.com/ethereum-optimism/optimism/pull/15615#discussion_r2071998790) please.